test/verify/run-tests

```diff
@@ -21,32 +82,124 @@ def check_valid(filename):
+    # Build the list of tests we'll parallellize and the ones we'll run serially
+    test_loader = unittest.TestLoader()
+    parallel_tests = []
+    serial_tests = []
+    test_id = 1
+    result = 0
+    jobs = 1 if opts.list else opts.jobs
+
     for filename in glob.glob(os.path.join(os.path.dirname(__file__), "check-*")):
         name = check_valid(filename)
         if not name or not os.path.isfile(filename):
             continue
-        with open(filename, 'rb') as fp:
-            module = imp.load_module(name, fp, filename, ("", "rb", imp.PY_SOURCE))
-            suite.addTest(loader.loadTestsFromModule(module))
+        loader = importlib.machinery.SourceFileLoader(name, filename)
+        module = importlib.util.module_from_spec(importlib.util.spec_from_loader(loader.name, loader))
+        loader.exec_module(module)
+        for test_suite in test_loader.loadTestsFromModule(module):
+            for test in test_suite:
+                test_method = getattr(test.__class__, test._testMethodName)
+                test_str = "{0}.{1}".format(test.__class__.__name__, test._testMethodName)
+                if opts.tests and not any([t in test_str for t in opts.tests]):
+                    continue
+                if getattr(test_method, "_testlib__non_destructive", False):
+                    serial_tests.append(Test(test_id, build_command(filename, test_str, opts)))
+                else:
+                    parallel_tests.append(Test(test_id, build_command(filename, test_str, opts)))
+                test_id += 1
+
+    print("1..{0}".format(len(parallel_tests) + len(serial_tests)))
+    sys.stdout.flush()
+
+    if serial_tests:
+        testlib.MachineCase.get_global_machine()
+    for test in serial_tests:
+        test.command.insert(-2, "--machine")
+        test.command.insert(-2, "{0}:{1}".format(testlib.MachineCase.get_global_machine().ssh_address,
+                                                 testlib.MachineCase.get_global_machine().ssh_port))
+        test.command.insert(-2, "--browser")
+        test.command.insert(-2, "{0}:{1}".format(testlib.MachineCase.get_global_machine().web_address,
+                                                 testlib.MachineCase.get_global_machine().web_port))
+
+    running_tests = []
+    serial_test = None
+    while serial_tests or parallel_tests or running_tests:
+        if len(running_tests) < jobs:
+            if serial_tests and serial_test is None:
+                test = serial_tests.pop(0)
+                test.process = subprocess.Popen(test.command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+                fcntl.fcntl(test.process.stdout.fileno(), fcntl.F_SETFL, os.O_NONBLOCK)
+                running_tests.append(test)
+                serial_test = test
+            elif parallel_tests:
+                test = parallel_tests.pop(0)
+                test.process = subprocess.Popen(test.command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```

Comment:
DRY.

also: you shouldn't call `F_SETFL` like that.  the usual paradigm is to call `F_GETFL` first, modify the bitmask and then call `F_SETFL` on the result.

Reply:
What's DRY?